sale_delivery_rate/models/delivery_carrier.py

```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 LasLabs Inc.
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from openerp import api, fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = 'delivery.carrier'
+
+    delivery_type = fields.Selection(
+        selection_add=[('base_on_rate', 'Based on Rates')],
+    )
+
+    @api.model
+    def base_on_rate_get_shipping_price_from_so(self, sale_orders):
+        prices = []
+
+        for sale_order in sale_orders:
+            order_rate = self.env['delivery.carrier.rate'].search([
```

Comment:
I'd add a constraint in `delivery.carrier.rate` for `UNIQUE(sale_order_id, service_id)`